jaeger-core/src/main/java/com/uber/jaeger/Tracer.java

```diff
@@ -129,90 +90,50 @@ void reportSpan(Span span) {
-        <T> ExtractorFactory<T> getExtractor(Class<?> carrierType) {
-            return (ExtractorFactory<T>) lookup(carrierType, extractors, "extractor");
-        }
-
-        public <T> Injector<T> register(Class<T> carrierType, Injector<?> injector) {
-            if (injectors.containsKey(carrierType)) {
-                return null;
-            }
-
-            // We store injectors with multiple carrierTypes, but only return the injector for the carrierType
-            // that we just registered.  Thus this is type unsafe.  However logically we know that the Injector
-            // type return the type we just associated with its key.
-            @SuppressWarnings("unchecked")
-            Injector<T> receivedInjector =  (Injector<T>) injectors.put(carrierType, injector);
-            return receivedInjector;
-        }
-
-        public <T> ExtractorFactory register(Class<T> carrierType, ExtractorFactory extractor) {
-            if (extractors.containsKey(carrierType)) {
-                return null;
-            }
-
-            return extractors.put(carrierType, extractor);
+    public <T> io.opentracing.SpanContext extract(Format<T> format, T carrier) {
+        Extractor<T> extractor = registry.getExtractor(format);
+        if (extractor == null) {
+            throw new UnsupportedFormatException(format);
         }
+        return extractor.extract(carrier);
     }
 
-    private abstract class AbstractSpanBuilder implements io.opentracing.Tracer.SpanBuilder {
+    private class SpanBuilder implements io.opentracing.Tracer.SpanBuilder {
 
         private String operationName = null;
         private long start;
+        private SpanContext firstParent;
         private final Map<String, Object> tags = new HashMap<>();
 
-        protected abstract Map<String, String> getBaggage();
+        SpanBuilder(String operationName) {
+            this.operationName = operationName;
+        }
 
-        protected abstract TraceContext getContext();
+        @Override
+        public io.opentracing.Tracer.SpanBuilder asChildOf(io.opentracing.SpanContext parent) {
+            return addReference(References.CHILD_OF, parent);
+        }
 
-        AbstractSpanBuilder(String operationName) {
-            this.operationName = operationName;
+        @Override
+        public io.opentracing.Tracer.SpanBuilder asChildOf(io.opentracing.Span parent) {
+            return addReference(References.CHILD_OF, parent.context());
         }
 
         @Override
-        public io.opentracing.Tracer.SpanBuilder withOperationName(String operationName) {
-            this.operationName = operationName;
+        public io.opentracing.Tracer.SpanBuilder addReference(String referenceType, io.opentracing.SpanContext referencedContext) {
+            if (firstParent == null && (
+                    referenceType == References.CHILD_OF || referenceType == References.FOLLOWS_FROM)) {
+                this.firstParent = (SpanContext) referencedContext;
```

Comment:
this is how the OT API is designed, when you getting a Span (or SpanContext) from the Tracer, you're getting a top-level interface, and subsequent operations inside the tracer must downcast it to the original impl. The only way to avoid that is to use generics, which will require all end-user code to be templatized on the type of tracer/span/spanCtx, which is not feasible.